### PR TITLE
Memory scheduling deadlock

### PR DESCRIPTION
**Pull request by jaewan**
## Why are these changes needed?

Implementation of Deadlock Detection1.
This checks the number of leased_workers and spinning_tasks(blocked tasks from triggering spill)
When the numbers match, consider it as a deadlock and trigger spill.

### CODE REVIEW

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
Two things:
1. Ideally, we shouldn't leak the number of spilled workers to the plasma request queue. Instead, can we do this logic in the task manager? The reason is that we already has a way for the task manager to notify the request queue when to block/allow spill, so we can build off of this codepath instead of creating a new one.
2. Can you add a test in Python that demonstrates the new behavior? We want one that previously did not work when blocking spills was enabled, and now it does work. Also, we should compare the runtime to the default Ray implementation that always spills.

**Comment by jaewan** [comment]
> Two things:
> 
> 1. Ideally, we shouldn't leak the number of spilled workers to the plasma request queue. Instead, can we do this logic in the task manager? The reason is that we already has a way for the task manager to notify the request queue when to block/allow spill, so we can build off of this codepath instead of creating a new one.
> 2. Can you add a test in Python that demonstrates the new behavior? We want one that previously did not work when blocking spills was enabled, and now it does work. Also, we should compare the runtime to the default Ray implementation that always spills.

1. I changed the implementation to trigger spill from task_manager. This is done by setting should_spill_ in object store via io_post. 
    !!! Can you check this Stephanie? is this design OK or should I get back to previous design?
    1.1 I also changed the setting should_spill_ from task_manager happens only when it is true (1. When #spinning_workers==#leased_workers, 2. When evict_tasks() called but did not evict anything). In previous design should_spill_ is always set after calling evict_tasks() and block_spill(). So no need for resetting should_spill_ from the object store. Advantage of this is since the calls are asynchronous, there's a time difference between setting should_spill_ from the task_manager to object_store. More spill() will be trigger in the meantime. But I thought it's wasting communication cost so changed the implementation. When should_spill_ is set, object_manager triggers spill once and set it off so task_manager doesn't need to reset it. 

2. We have a separate PR for test scripts. Do you want me to add test of this PR here or to the test script PR?

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 1]
```diff
@@ -196,19 +202,32 @@ Status CreateRequestQueue::ProcessRequests() {
...
+	  if (enable_blocktasks || enable_evicttasks || enable_blocktasks_spill) {
         RAY_LOG(DEBUG) << "[JAE_DEBUG] calling object_creation_blocked_callback (" 
 			<< enable_blocktasks <<" " << enable_evicttasks << " " 
 			<< enable_blocktasks_spill << ") on priority "
 			<< lowest_pri;
-		if(!block_tasks_required && !evict_tasks_required){
-	      on_object_creation_blocked_callback_(lowest_pri, enable_blocktasks , enable_evicttasks);
+		if(enable_blocktasks_spill){
+          RAY_LOG(DEBUG) << "[JAE_DEBUG] task " << task_id << " spins"; 
+		  for(auto it = queue_.begin(); it != queue_.end(); it++){
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/create_request_queue.cc:212]: Can't we just use queue.size()?

**Comment by stephanie-wang** [comment]
The implementation looks much better!

> We have a separate PR for test scripts. Do you want me to add test of this PR here or to the test script PR?

Can you add just one script that previously deadlocked and now works with this PR?

**Pull request merged**